Add option for ALAMO model as expression

eslickj opened the pull request:
This adds an option so that when you build a Pyomo model from an ALAMO surrogate, you have the option to build it as expressions rather than constraints.

The surrogate tests are also very delicate.  The Pysmo tests included tests that had multiple equally good results and a lot of tests for both Pysomo and ALAMO depend on comparing floating point numbers as strings.  I have a few Pysmo tests for save/load pickled results that still fail for me, but if the CI tests pass, we can investigate that later. 

This may partly address issue #841

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

andrewlee94 commented on idaes/core/surrogate/alamopy.py:
```diff
@@ -1226,25 +1226,43 @@ def evaluate_surrogate(self, inputs):
             block: Pyomo Block component to be populated with constraints.
-            additional_options: None
-               No additional options are required for this surrogate object
+            additional_options (dict): as_expression (bool) is an optional
```
1. Doc string does not match argument.
2. Why make this a dict of options rather than a list of separate arguments?

eslickj replied:
I didn't write it, it was already there and already a dict.  I think the dict comes from keyword args to ``build_model`` in the general class.  This isn't something users use directly.  I'll fix the doc string, but I'm not really sure how to properly document it.

I suppose the right way would be to add it to the build_model doc string, but it only works with ALAMO for now.

andrewlee94 replied:
Is this method called indirectly via another method on the actual SurrogateModel block (i.e. `build_model`)?

andrewlee94 reviewed (changes requested):
A couple of comments on the code, plus a few more general things:

1. Can you comment on the usage of the as_expressions option, and how that would actually get used in practice?
2. The new options need documentation, and probably need to comment on how they should be used.

andrewlee94 reviewed (approved):
Looks OK to me, pending getting the Pysmo tests fixed.

eslickj commented:
Ok, I updated the tests for Pysmo saves to save then load to a new object and compare the two.  I think depending on strings to compare floating point results should be okay if you always stick to the same platform, but it's pretty delicate. 

I added a data point to the Pysmo tests, which should make the solution unique now.  I think that should fix #841.

andrewlee94 commented on idaes/core/surrogate/tests/test_pysmo_surrogate.py:
```diff
@@ -1373,9 +1373,13 @@ def test_populate_block_unisurrogate_poly(self, pysmo_surr1):
+        cstr = cstr.replace("inputs[x1]", "1")
+        cstr = cstr.replace("inputs[x2]", "5")
+        assert eval(cstr) == pytest.approx(0, abs=1e-8)
+        # assert str(blk.pysmo_constraint["z1"].body) == (
```
Is this commented code still necessary?

eslickj replied:
No, I just forgot to delete the old code.  Will remove.

OOAmusat commented on idaes/core/surrogate/tests/test_pysmo_surrogate.py:
```diff
@@ -1373,9 +1371,10 @@ def test_populate_block_unisurrogate_poly(self, pysmo_surr1):
-        assert str(blk.pysmo_constraint["z1"].body) == (
-            "outputs[z1] - (-75.26111111111476 - 8.815277777775934*inputs[x1] + 18.81527777777826*inputs[x2] - 2.2556956302821618e-13*(inputs[x2]*inputs[x1]))"
-        )
+        cstr = str(blk.pysmo_constraint["z1"].body).replace("outputs[z1]", "10")
```
@eslickj If I understand this correctly you're testing the value of the evaluated expression, not the expression itself? The purpose of this test was to ensure that the pyomo block was actually generating the correct/accurate constraint on the block. There is are separate tests for evaluating the surrogate(s).

eslickj replied:
The issue is that the test compares a string, so the floating point number have to be exactly the same.  On arm64 processors at the very least there are slight differences.  The only way I can really think to test this is to parse our the coefficients from the string and compare.  It seemed like a lot of work.

eslickj replied:
I guess this shouldn't be too hard with regular expressions.

OOAmusat replied:
You're right, it probably isn't worth the effort. We can go with that you have.

OOAmusat reviewed (approved):
LGTM
